core/src/main/java/org/wildfly/plugin/core/Deployment.java

```diff
@@ -0,0 +1,299 @@
+     */
+    public Deployment(final File content) {
+        this(Assertions.requiresNotNullParameter(content, "content").toPath());
+    }
+
+    /**
+     * Creates a new deployment for the path. If the path is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the path containing the content
+     */
+    public Deployment(final Path content) {
+        this.pathContent = Assertions.requiresNotNullParameter(content, "content");
+        inputStreamContent = null;
+        serverGroups = new LinkedHashSet<>();
+    }
+
+    /**
+     * Creates a new deployment for the input stream. The name is required for the deployment and cannot be {@code null}.
+     * If {@link #setName(String)} with a {@code null} argument is invoked when using this constructor an
+     * {@link IllegalArgumentException} will be thrown.
+     *
+     * @param content the input stream representing the content
+     * @param name    the name for the deployment
+     */
+    public Deployment(final InputStream content, final String name) {
+        this.pathContent = null;
+        inputStreamContent = Assertions.requiresNotNullParameter(content, "content");
+        serverGroups = new LinkedHashSet<>();
+        this.name = Assertions.requiresNotNullParameter(name, "name");
+    }
+
+    /**
+     * Creates a new deployment for the file. If the file is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the file containing the content
+     *
+     * @return the deployment
+     */
+    public static Deployment of(final File content) {
+        return new Deployment(content);
+    }
+
+    /**
+     * Creates a new deployment for the path. If the path is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the path containing the content
+     *
+     * @return the deployment
+     */
+    public static Deployment of(final Path content) {
+        return new Deployment(content);
+    }
+
+    /**
+     * Creates a new deployment for the input stream. The name is required for the deployment and cannot be {@code null}.
+     * If {@link #setName(String)} with a {@code null} argument is invoked when using this static factory an
+     * {@link IllegalArgumentException} will be thrown.
+     *
+     * @param content the input stream representing the content
+     * @param name    the name for the deployment
+     *
+     * @return the deployment
+     */
+    public static Deployment of(final InputStream content, final String name) {
+        return new Deployment(content, name);
+    }
+
+    /**
+     * Adds a server group for the deployment.
+     *
+     * @param serverGroup the server group to add
+     *
+     * @return this deployment
+     */
+    public Deployment addServerGroup(final String serverGroup) {
+        serverGroups.add(serverGroup);
```

Comment:
I don't see this being used in a multi-threaded environment often. The only situation where I could see it being useful is if you're deploying the same content to several servers at once. I will document that this is not thread-safe however.

Also if it becomes immutable it would require either a lot of constructors or a builder. I'm not real keen on making yet another builder for an API that's really supposed to be pretty simple and here for convenience.